### ISSUE TITLE
Improve range sync invalid batch handling

### ISSUE THREAD

**Issue by AgeManning** [sigp/lighthouse#970]
## Description

The range sync downloads batches of blocks from a pool of peers. A previously completed batch is cached until a newer batch can be successfully processed. This is because a successfully processed batch may contain no-blocks or the end of the batch may-contain all skip slots. Alternatively a lazy-peer may only return half a batch. We can identify this, by ensuring that at least one block in the following batch succeeds. 

Currently, due the to the block processor design, we return a boolean as to whether the entire batch succeeds. We therefore do not know if a single block succeeded, in which case we could be sure all previous batches are valid and we only need to download the current failed batch. 

The current version re-downloads previous cached batches as we have not verified them and this is potentially unnecessary. 

If we update the block processor to return extra information indicating if at least the first block was successfully downloaded, we could verify previous cached batches and only download the current;y failed batch.

**Comment by paulhauner** [comment]
> If we update the block processor to return extra information indicating if at least the first block was successfully downloaded, we could verify previous cached batches and only download the current;y failed batch.

We'll have to pass this information down from [`BeaconChain::process_chain_segment`](https://github.com/sigp/lighthouse/blob/93da74e7d630602efc6ccbf731128b90e28ae3a1/beacon_node/beacon_chain/src/beacon_chain.rs#L1148) too. Perhaps we need to return something odd like `(Vec<Hash256>, Result<(), BlockError>)` so that we can detect if any blocks were imported... We can think more about this once someone picks up the issue :)

**Comment by AgeManning** [comment]
Agreed. I think @divagant-martian was picking up sync-related issues like this. So potentially she's interested in this also.

**Comment by divagant-martian** [comment]
I'll take it. Thanks :)

**Comment by paulhauner** [comment]
So, AUIU the problem is that `BeaconChain::process_chain_segement` returns a `Result` which only gives you two pieces of information:

1. `Ok(..)`: all of the blocks were valid and successfully imported.
2. `Err(..)`: one or more of the blocks were invalid and zero or more blocks were imported.

AIUI, you need a third piece of information that indicates: "one or more of the blocks were valid and successfully imported".

I think that the best solution here would be to define this struct:

```rust
pub struct ChainSegmentError {
  blocks_imported: Vec<Hash256>,
  error: BlockError
}
```

Then, we change the `process_chain_segment` signature to:

```rust
   pub fn process_chain_segment(
        &self,
        chain_segment: Vec<SignedBeaconBlock<T::EthSpec>>,
    ) -> Result<Vec<Hash256>, ChainSegmentError>
```

Then you could have the following logic:

```rust
match chain.process_chain_segment(..) {
  Ok(roots) if roots.is_empty() => // We didn't validate or import any blocks.
  Ok(roots) => // We imported some blocks.
  Err(e) if e.roots.is_empty() => // No blocks were imported to to invalid block or error.
  Err(e) => //  Some blocks were imported, but not all due to an invalid block or error.
}
```
This will make returning errors in `process_chain_segment` a little weird, but it's the most ergonomic solution I can think of. It'll also mean some re-jigging of the tests, but it should be straight-forward. 

Open to suggestions ofc :)

**Comment by divagant-martian** [comment]
From @paulhauner's comment I take this belongs to v0.2.0 but from @AgeManning's it would seem to belong to master?

**Comment by paulhauner** [comment]
> From @paulhauner's comment I take this belongs to v0.2.0 but from @AgeManning's it would seem to belong to master?

@AgeManning is the authority here, I'll let him respond :)

**Comment by AgeManning** [comment]
I'm merging master into v0.2.0 today, so it will be one and the same :)

**Comment by AgeManning** [comment]
Resolved in #996
